### PR TITLE
Intersphinx, refactoring and intersphinx_disable_domains

### LINKED ISSUE #2068: Add an option to not use intersphinx references as a fallback
That intersphinx references are used as fallbacks if a reference is not available in a project recently caused us a problem.

We have a site (`latest`) which is the latest build of `master`.

We use intersphinx as follows, so we can link to the latest documentation in some sections:

``` python
intersphinx_mapping = {
    'latest': ('http://doc-dev.clusterhq.com/', None),
}
```

``` rst
Make sure to follow the :ref:`latest documentation <latest:release-process>` when doing a release.
```

Our docs included a label (`.. foo:`) so this reference was available on `latest`.
This label was removed on a branch, along with one of the two references to this label.
In this case we would expect that building the branch would fail, as there is a reference to a label which doesn't exist.
Unexpectedly, the branch built successfully, because the label was found in `latest` as a fall back when it was not found locally. The branch was merged, and then later, when `latest` changed (because `master` was built again) builds stopped working because there was a reference to a non-existent label.

It would be good to have an option to not fall back, maybe something like `nitpicky`.

### PR DESCRIPTION
**Pull request by jakobandersen**
### Feature or Bugfix
- Feature
- Bugfix
- Refactoring

### Purpose
In short, this PR implements https://github.com/sphinx-doc/sphinx/pull/8981#issuecomment-879815795 and thus fixes #2068 and replaces #8981 (second commit) and part of #8929 (first commit).
At the same time the PR serves as a basis for a better implementation of #9062, and contains part of the refactoring in #8929.

### Detail
The first commit is primarily refactoring, though in the case where a target contains a prefix ``something:`` and the resolution fails, the old code would strip the prefix in the final output. Now this prefix is shown in the output to avoid hiding the information from the source code. I consider this a minor "bug" fix.

The refactoring consists splitting the resolution into multiple functions:
- The top-level is the three functions
  - ``resolve_reference_in_inventory``: lookup in a user-specified inventory
  - ``resolve_reference_any_inventory``: lookup in any inventory
  - ``resolve_reference_detect_inventory``: try first ``any`` and otherwise try to extract a prefix from the target. The ``missing_reference`` handler directly calls this one.
- The domain-specific resolution is in ``_resolve_reference_in_domain`` and ``_resolve_reference_in_domain_by_target``. Abstractly, these are the ones that need to be split and moved to the domain classes (see #8929).

The second commit implements ``intersphinx_disable_domains`` which defaults to ``[]`` for backwards compatibility. For now I added
```py
intersphinx_disabled_domains = ['std']
```
to the ``conf.py`` generated by ``sphinx-quickstart``. The rationale is that I believe ``:ref`` and ``:doc:`` references most often are intended to be local, and a simple typo or missing file may silently make them resolve to external documents (i.e., the problem discussed in #2068 and #8981. All other built-in domains either do not export objects for intersphinx (``changeset``, ``citation``, ``index``, ``math``) or they are programming language domains (``c``, ``cpp``, ``js``, ``py``, ``rst``) which generates cross-references in declarations where the user can not add an inventory specification.
Note, ``std`` also contains the object types ``term``, ``token``, ``envvar``, and ``cmdoption``, but my guess is that users rarely have external links for these, and in all cases one can provide an explicit inventory specification.

### CODE REVIEW

**Comment by jakobandersen** [comment]
@adamtheturtle, @nijel, and @phlax, does this solve your original issues?

**Comment by adamtheturtle** [comment]
@jakobandersen I'm no longer on the project with the issue so I can't really say.

**Comment by phlax** [comment]
> @adamtheturtle, @nijel, and @phlax, does this solve your original issues?

i can test

**Comment by nijel** [comment]
It seems that it should address the issue I had. Unfortunately, I probably won't have time to test this in next two weeks...

**Comment by astrojuanlu** [comment]
I just tested `intersphinx_disabled_domains = ['std']` and it works as expected. Previously, `` :doc:`relative <contents>` `` would look a `contents` in any of the intersphinx-mapped projects if it happened to not exist in that relative path, now it gives a `WARNING: unknown document: contents`. On the other hand, I can still do `` :doc:`Sphinx contents <sphinx:contents>` `` and it will correctly point to https://www.sphinx-doc.org/en/master/contents.html.

The name `intersphinx_disabled_domains` seems to suggest that it's going to disable the `:doc:` resolution altogether though. The documentation clarifies it (emphasis mine):

> When a cross-reference **without an explicit inventory specification** is being resolve [sic] by intersphinx, skip resolution if [...] the domain of the cross-reference is in this list [...]

But I can't think of a better name that is not super long (`intersphinx_disabled_automatic_fallback_domains`).

Some more comments on your (super thorough!) pull request overview:

> All other built-in domains either do not export objects for intersphinx (`changeset`, `citation`, `index`, `math`)

- To my understanding, `changeset`, `citation`, and `index` are not documented in https://www.sphinx-doc.org/en/master/usage/restructuredtext/domains.html, right? Should they?
- Would it be interesting for `math` to export objects for intersphinx?

> or they are programming language domains (`c`, `cpp`, `js`, `py`, `rst`) which generates cross-references in declarations where the user can not add an inventory specification

I'm not sure I follow. The `py` domain _does_ allow the user to add an inventory specification, right? I just tried `` :py:class:`sphinx:sphinx.builders.changes.ChangesBuilder` `` and it works fine. However, you mentioned in https://github.com/sphinx-doc/sphinx/pull/9424#discussion_r666948453 that "the roles in the C and C++ domains do not support this prefix".

**Comment by jakobandersen** [comment]
> > is being resolve [sic] by intersphinx

(Typo fixed)

> But I can't think of a better name that is not super long (`intersphinx_disabled_automatic_fallback_domains`).

Right, I got to the same conclusion. I was thinking of trying to get "implicit" in there, but did find a good name.

> > All other built-in domains either do not export objects for intersphinx (`changeset`, `citation`, `index`, `math`)
> 
> * To my understanding, `changeset`, `citation`, and `index` are not documented in https://www.sphinx-doc.org/en/master/usage/restructuredtext/domains.html, right? Should they?

Hmm, like the standard domain, I guess they contain roles/directives that are not naturally described in terms of a domain. I don't know the details of them, and haven't used them my self, but some bits I could find:
- for ``changeset`` it has the directives at https://github.com/sphinx-doc/sphinx/blob/810a1e2988b14f4d139b5ef328a91967f5ed7a08/sphinx/domains/changeset.py#L155-L157, and it seems to interact with the ``changes`` builder.
  Docs: https://www.sphinx-doc.org/en/master/usage/restructuredtext/directives.html#directive-versionadded
- ``citation`' seems to handle the ``citation`` node from docutils (https://docutils.sourceforge.io/docs/ref/doctree.html#citation)
- the directive/role in ``index`` seems to have docs at https://www.sphinx-doc.org/en/master/usage/restructuredtext/directives.html#directive-index

> * Would it be interesting for `math` to export objects for intersphinx?

Yes, it looks like the math domain seems to contain equation references already for internal resolution, so I don't immediately see why they couldn't be exposed via intersphinx. I guess no user has requested it before?

> > or they are programming language domains (`c`, `cpp`, `js`, `py`, `rst`) which generates cross-references in declarations where the user can not add an inventory specification
> 
> I'm not sure I follow. The `py` domain _does_ allow the user to add an inventory specification, right? I just tried `` :py:class:`sphinx:sphinx.builders.changes.ChangesBuilder` `` and it works fine. However, you mentioned in [#9424 (comment)](https://github.com/sphinx-doc/sphinx/pull/9424#discussion_r666948453) that "the roles in the C and C++ domains do not support this prefix".

For references generated from roles intended to reference a single entity, yes, there the Python domain (and perhaps js and rst?) support the prefix to some degree. However, this is because they either don't need to perform detailed parsing of the role argument, or they specifically need to take intersphinx into account. This is why the prefix in a C/C++ role gives a parsing warning/error (#8418).
The C and C++ domains additionally support a role where you can render an (almost) arbitrary type or expression (``:expr:`` role), where it would be strange to have intersphinx prefixes scattered throughout.
In general, all the programming domains generally generate cross-references implicitly in declarations, e.g., in Python:
```rst
.. py:function:: f(i: int) -> dict
```
Here ``int`` and ``dict`` are cross-references that will be resolved as normal, including with intersphinx, and I don't think there is a way to annotate with which inventory to perform lookup (and I don't think there should be).

**Comment by astrojuanlu** [comment]
Thanks a lot for the clarifications about these extra roles and directives. I am still trying to make sense of the Sphinx API and how it is described in the documentation.

> Yes, it looks like the math domain seems to contain equation references already for internal resolution, so I don't immediately see why they couldn't be exposed via intersphinx. I guess no user has requested it before?

Done :) #9483

> However, this is because they either don't need to perform detailed parsing of the role argument, or they specifically need to take intersphinx into account.

Thanks! That's what I understood from your previous comments. I was distracted by your wording here, I now get that programming-language domains don't _necessarily_ or _always_  support explicit inventory specification (aka intersphinx prefixes).

FWIW, this looks like a good addition to me and I'm looking forward for this to pave the way towards #9062.

File: sphinx/templates/quickstart/conf.py_t [hunk 1]
```diff
@@ -108,6 +108,10 @@ html_static_path = ['{{ dot }}static']
 intersphinx_mapping = {
     'python': ('https://docs.python.org/3', None),
 }
+
+# Prevent accidental intersphinx resolution for labels, documents, and other
+# basic cross-references.
+intersphinx_disabled_domains = ['std']
```
Comment (tk0miya) [inline sphinx/templates/quickstart/conf.py_t]: I think this should be set as a default setting of `intersphinx_disabled_domains` instead of the generated conf.py. I guess you tried not to introduce a breaking change. But this implementation is not good for me.

File: doc/usage/extensions/intersphinx.rst [hunk 2]
```diff
@@ -148,6 +148,25 @@ linking:
...
+   .. versionadded:: 4.2
+
+   A list of strings being the name of a domain, or the special name ``all``.
+   When a cross-reference without an explicit inventory specification is being
+   resolved by intersphinx, skip resolution if either the domain of the
+   cross-reference is in this list or the special name ``all`` is in the list.
+
+   For example, with ``intersphinx_disabled_domains = ['std']`` a cross-reference
+   ``:doc:`installation``` will not be attempted to be resolved by intersphinx, but
+   ``:doc:`otherbook:installation``` will be attempted to be resolved in the
```
Comment (tk0miya) [inline doc/usage/extensions/intersphinx.rst]: I think explicit references should be also disabled. It would be replaced by https://github.com/sphinx-doc/sphinx/pull/9062#issuecomment-879861369.

File: sphinx/templates/quickstart/conf.py_t (hunk 1 above)
Comment (jakobandersen) [reply sphinx/templates/quickstart/conf.py_t]: I agree, and indeed I wanted to avoid breaking changes to get the functionality released earlier.  My idea was to make a another PR after this, but against master, that changes the default value and removes this bit again. Would that be fine?
Alternatively I can also just remove this now and still make the default value change in master?

File: doc/usage/extensions/intersphinx.rst (hunk 2 above)
Comment (jakobandersen) [reply doc/usage/extensions/intersphinx.rst]: Note that I wrote this PR as if https://github.com/sphinx-doc/sphinx/pull/9062#issuecomment-879861369 didn't exist. My plan is to have the functionality change incrementally:
1. This PR, for which this part is as intended: the user must intentionally write something intersphinx-specific to activate it for the disabled domains.
2. Then an updated #9062 which uses the functionality in this PR. By then the old style of explicit intersphinx references, as the one you quoted here, can be deprecated/removed.
3. The whole domain delegation part.

So in short, yes, once an updated #9062 has been merged, I don't mind this option disabling the explicit ones as well, but I don't think it is harmful if they are not.

**Comment by phlax** [comment]
@jakobandersen i tested this branch this morning (here https://github.com/envoyproxy/envoy/pull/18090) and afaict everything is working as i would hope with non/intersphinx ref links - thanks for addressing this

lmk if you would like any further testing

**Comment by jakobandersen** [comment]
@phlax, thanks for the testing!

File: sphinx/templates/quickstart/conf.py_t (hunk 1 above)
Comment (jakobandersen) [reply sphinx/templates/quickstart/conf.py_t]: Quickstart stuff removed.

**Comment by jakobandersen** [comment]
@tk0miya, rebased and quickstart stuff removed. I think this is good to merge, assuming something like the following incremental roadmap:

1. ~(4.x) this PR.~
2. ~(master) change the default value of ``intersphinx_disabled_domains`` to ``['std']`` as a breaking change.~ (see #9804)
3. (4.x) merge #9062 updated with the following:
   - Use the refactored interface of this PR.
   - Change the documentation to not refer to the old explicit inventory syntax (e.g.,  ``:doc:`otherbook:installation` ``) but instead use the new role. This includes modifying the docs added in this PR.
   - Optionally: let ``intersphinx_disabled_domains`` apply to the old explicit inventory syntax as well, not that the PR provides an alternative.
4. (master) deprecate the old explicit inventory syntax.
5. (master) The whole domain delegation part and other large changes.

**Comment by tk0miya** [comment]
Note: I found https://github.com/sphinx-doc/sphinx/pull/9626 contains remote `:envvar:` references. The environment variables belong to the std domain, but this disables the reference by default. I need to reconsider the target of this feature.

**Comment by jakobandersen** [comment]
> Note: I found #9626 contains remote `:envvar:` references. The environment variables belong to the std domain, but this disables the reference by default. I need to reconsider the target of this feature.

Doh, right. Disabling all of ``std``is a no-go. Theoretically we could split ``std`` into two or more domains, with one of them having ``doc``, but realistically this might break a lot of things if we are not careful.
As an alternative, I have pushed the following changes:
- Instead of ``intersphinx_disabled_domains`` it is now ``intersphinx_disabled_refs``. Not sure if this is the best name.
- You can write ``all`` and domain names as before, but now also specific references types, e.g., ``std:doc``.

The breaking change to the master branch after this PR is then to default this option to ``['std:doc']``.

**Comment by astrojuanlu** [comment]
Anything we can do to move this forward? It would be great to have this in Sphinx 4.3, I could play with it a bit and document it as part of the tutorial.

File: CHANGES [hunk 3]
```diff
@@ -47,6 +47,14 @@ Features added
 * #9695: More CSS classes on Javascript domain descriptions
 * #9683: Revert the removal of ``add_stylesheet()`` API.  It will be kept until
   the Sphinx-6.0 release
+* #2068, add :confval:`intersphinx_disabled_domains` for disabling
```
Comment (tk0miya) [inline CHANGES]: Does this configuration still exist? It seems already replaced by `intersphinx_disabled_refs ` on discussing.

File: doc/usage/extensions/intersphinx.rst [hunk 4]
```diff
@@ -148,6 +148,32 @@ linking:
       exception is raised if the server has not issued a response for timeout
       seconds.
 
+.. confval:: intersphinx_disabled_refs
+
+   .. versionadded:: 4.3
+
+   A list of strings being either:
+
+   - the name of a specific reference type,
```
Comment (tk0miya) [inline doc/usage/extensions/intersphinx.rst]: As this sentence describes, it would be better to rename `intersphinx_disabled_reftypes`. Internally, these items are called as "refdomain" and "reftype". So it's easy to understand from the implementation sight.

File: doc/usage/extensions/intersphinx.rst [hunk 5]
```diff
@@ -148,6 +148,32 @@ linking:
...
+   For example, with ``intersphinx_disabled_refs = ['std:doc']``
+   a cross-reference ``:doc:`installation``` will not be attempted to be
+   resolved by intersphinx, but ``:doc:`otherbook:installation``` will be
+   attempted to be resolved in the inventory named ``otherbook`` in
+   :confval:`intersphinx_mapping`.
+   At the same time, all cross-references generated in, e.g., Python,
+   declarations will still be attempted to be resolved by intersphinx.
+
+   If ``all`` is in the list of domains, then no references without an explicit
+   inventory will be resolved by intersphinx.
```
Comment (tk0miya) [inline doc/usage/extensions/intersphinx.rst:178]: We've used `*` to represent "all" in some configuration. So it might be better to use it instead.

File: doc/usage/extensions/intersphinx.rst [hunk 6]
```diff
@@ -148,6 +148,32 @@ linking:
       exception is raised if the server has not issued a response for timeout
       seconds.
 
+.. confval:: intersphinx_disabled_refs
```
Comment (tk0miya) [inline doc/usage/extensions/intersphinx.rst]: There is no explanation for the default of this setting. Could you add it?

File: sphinx/util/typing.py [hunk 7]
```diff
@@ -70,7 +70,8 @@ def _evaluate(self, globalns: Dict, localns: Dict) -> Any:
 TitleGetter = Callable[[nodes.Node], str]
 
 # inventory data on memory
-Inventory = Dict[str, Dict[str, Tuple[str, str, str, str]]]
+InventoryInner = Tuple[str, str, str, str]
```
Comment (tk0miya) [inline sphinx/util/typing.py]: I prefer to call this as "item" or "entry". What do you think?

File: sphinx/ext/intersphinx.py [hunk 8]
```diff
@@ -258,105 +259,212 @@ def load_mappings(app: Sphinx) -> None:
...
+
+    Resolution is tried with the target as is in any inventory.
+    """
+    return _resolve_reference(env, None, InventoryAdapter(env).main_inventory,
+                              honor_disabled_refs,
+                              node, contnode)
+
+
+def resolve_reference_detect_inventory(env: BuildEnvironment,
+                                       honor_disabled_refs: bool,
```
Comment (tk0miya) [inline sphinx/ext/intersphinx.py]: Why this function and `resolve_reference_any_inventory()` take `honor_disabled_refs`? It seems the value is fixed to True. So no reason to take an argument like `resolve_reference_in_inventory()`.

File: doc/usage/extensions/intersphinx.rst [hunk 9]
```diff
@@ -148,6 +148,32 @@ linking:
...
+
+   .. versionadded:: 4.3
+
+   A list of strings being either:
+
+   - the name of a specific reference type,
+     e.g., ``std:doc``, ``py:func``, or ``cpp:class``,
+   - the name of a whole domain, e.g., ``std``, ``py``, or ``cpp``, or
+   - the special name ``all``.
```
Comment (tk0miya) [inline doc/usage/extensions/intersphinx.rst]: What do you think that this configuration always takes `{refdomain}:{reftype}` and wildcards (ex. `std:*`, `*`). Then this can be renamed to `intersphinx_disabled_reftypes`. It helps to simplify this configuration, I think.

File: sphinx/ext/intersphinx.py [hunk 10]
```diff
@@ -258,105 +259,212 @@ def load_mappings(app: Sphinx) -> None:
...
+        return None
+    return _resolve_reference_in_domain_by_target(inv_name, inventory, domain, objtypes,
+                                                  full_qualified_name, node, contnode)
+
+
+def _resolve_reference(env: BuildEnvironment, inv_name: Optional[str], inventory: Inventory,
+                       honor_disabled_refs: bool,
+                       node: pending_xref, contnode: TextElement) -> Optional[Element]:
+    # disabling should only be done if no inventory is given
+    honor_disabled_refs = honor_disabled_refs and inv_name is None
```
Comment (tk0miya) [inline sphinx/ext/intersphinx.py:364]: Is there any case that `honor_disabled_refs` is True, but `inv_name` is None?

**Comment by tk0miya** [comment]
LGTM with nits.

File: CHANGES (hunk 3 above)
Comment (jakobandersen) [reply CHANGES]: Good catch.

File: doc/usage/extensions/intersphinx.rst (hunk 5 above)
Comment (jakobandersen) [reply doc/usage/extensions/intersphinx.rst:178]: Good idea, changed to ``*``.

File: sphinx/util/typing.py (hunk 7 above)
Comment (jakobandersen) [reply sphinx/util/typing.py]: I agree, renamed to ``InventoryItem``.

File: sphinx/ext/intersphinx.py (hunk 10 above)
Comment (jakobandersen) [reply sphinx/ext/intersphinx.py:364]: Yes, that is the default for most ``pending_xref``:
1. From the ``missing_reference`` callback it will honor the disabling: https://github.com/jakobandersen/sphinx/blob/778a3fe6a7a4caf9152caa5c9b10d5378de671c2/sphinx/ext/intersphinx.py#L467
2. going through ``resolve_reference_detect_inventory``: https://github.com/jakobandersen/sphinx/blob/778a3fe6a7a4caf9152caa5c9b10d5378de671c2/sphinx/ext/intersphinx.py#L446
3. and then in ``resolve_reference_any_inventory`` the ``None`` is added: https://github.com/jakobandersen/sphinx/blob/778a3fe6a7a4caf9152caa5c9b10d5378de671c2/sphinx/ext/intersphinx.py#L428

File: sphinx/ext/intersphinx.py (hunk 8 above)
Comment (jakobandersen) [reply sphinx/ext/intersphinx.py]: Right, that parameter was pulled up to far. It's now hard-coded in this function. The new intersphinx role should use either ``resolve_reference_any_inventory()`` with ``honor_disabled_refs=False`` or ``resolve_reference_in_inventory()``, depending on if the user specifies an inventory.

File: doc/usage/extensions/intersphinx.rst (hunk 6 above)
Comment (jakobandersen) [reply doc/usage/extensions/intersphinx.rst]: Added.

File: doc/usage/extensions/intersphinx.rst (hunk 9 above)
Comment (jakobandersen) [reply doc/usage/extensions/intersphinx.rst]: I like that. Changed the format and the confvar name.

File: doc/usage/extensions/intersphinx.rst (hunk 4 above)
Comment (jakobandersen) [reply doc/usage/extensions/intersphinx.rst]: Indeed, renamed.

**Comment by jakobandersen** [comment]
> LGTM with nits.

Great, thanks! All nits addressed basically as suggested, so I'll merge once CI is green.

**Comment by tk0miya** [comment]
🎉
